File: app/src/main/java/com/productions/gizzmoo/pokemonpuzzleleague/settings/SettingsFragment.java
package com.productions.gizzmoo.pokemonpuzzleleague.settings;

import android.os.Bundle;
import android.preference.PreferenceFragment;

import com.productions.gizzmoo.pokemonpuzzleleague.R;

/**
 * Created by Chrystian on 4/9/2018.
 */

public class SettingsFragment extends PreferenceFragment {
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Load the preferences from an XML resource
        addPreferencesFromResource(R.xml.preferences);
    }

    @Override
    public void onPause() {
        getPreferenceManager().getSharedPreferences().unregisterOnSharedPreferenceChangeListener((PokemonPreference) findPreference("pref_pokemon_key"));
        super.onPause();
    }

    @Override
    public void onResume() {
        super.onResume();
        getPreferenceManager().getSharedPreferences().registerOnSharedPreferenceChangeListener((PokemonPreference) findPreference("pref_pokemon_key"));
    }
}

File: app/src/main/java/com/productions/gizzmoo/pokemonpuzzleleague/settings/SettingsFragment.kt
package com.productions.gizzmoo.pokemonpuzzleleague.settings

import android.os.Bundle
import androidx.fragment.app.DialogFragment
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import com.productions.gizzmoo.pokemonpuzzleleague.R
import android.graphics.drawable.ColorDrawable

class SettingsFragment : PreferenceFragmentCompat() {
    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        setDivider(ColorDrawable(resources.getColor(R.color.primary)))
        setDividerHeight(getDpiForDivider())
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.preferences, rootKey)
    }

    override fun onDisplayPreferenceDialog(preference: Preference) {
        when (preference) {
            is TrainerPreference -> showFragmentDialog(TrainerDialog.newInstance(preference))
            is PokemonPreference -> showFragmentDialog(PokemonDialog.newInstance(preference))
            is NumberPreference -> showFragmentDialog(NumberDialog.newInstance(preference))
            else -> super.onDisplayPreferenceDialog(preference)
        }
    }

    override fun onPause() {
        preferenceManager.sharedPreferences.unregisterOnSharedPreferenceChangeListener(findPreference("pref_pokemon_key") as PokemonPreference)
        super.onPause()
    }

    override fun onResume() {
        super.onResume()
        preferenceManager.sharedPreferences.registerOnSharedPreferenceChangeListener(findPreference("pref_pokemon_key") as PokemonPreference)
    }

    private fun showFragmentDialog(fragment: DialogFragment) {
        fragment.setTargetFragment(this, 0)
        fragment.show(fragmentManager, "android.support.v7.preference.PreferenceFragment.DIALOG")
    }

    private fun getDpiForDivider(): Int =
            (DIVIDER_SIZE * resources.displayMetrics.density + 0.5f).toInt()

    companion object {
        const val DIVIDER_SIZE = 1
    }
}